### PR TITLE
35) added spring security in project

### PR DESCRIPTION
**Pull request by AntonPanchishko**

### CODE REVIEW

File: src/main/java/com/dev/cinema/annotaion/ValidPassword.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    String field();
+
+    String fieldMatch();
+
+    Class<?>[] groups() default {};
+
+    Class<? extends Payload>[] payload() default {};
+
+    @Target({ ElementType.TYPE })
```
Comment (vicire) [inline src/main/java/com/dev/cinema/annotaion/ValidPassword.java]: We have only one class level validation, so we can remove this

File: src/main/java/com/dev/cinema/controller/CustomGlobalExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex,
+            HttpHeaders headers, HttpStatus status,
+            WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", new Date());
+        body.put("status", status.value());
+        List<String> errors = ex.getBindingResult()
+                .getFieldErrors()
```
Comment (vicire) [inline src/main/java/com/dev/cinema/controller/CustomGlobalExceptionHandler.java]: Better catch all errors, not only field

File: src/main/java/com/dev/cinema/controller/OrderController.java [hunk 3]
```diff
@@ -34,17 +35,26 @@ public OrderController(OrderService orderService,
...
     @GetMapping
-    public List<OrderResponseDto> getOrderHistory(@RequestParam Long userId) {
+    public List<OrderResponseDto> getOrderHistory(Authentication authentication) {
+        UserDetails details = (UserDetails) authentication.getPrincipal();
+        String email = details.getUsername();
         return orderService
-                .getOrdersHistory(userService.getById(userId))
+                .getOrdersHistory(userService
+                        .findByEmail(email)
+                        .orElseThrow(() -> new RuntimeException("Incorrect login or password")))
```
Comment (vicire) [inline src/main/java/com/dev/cinema/controller/OrderController.java]: At this moment we will get user, that 100% is in our database. So we can remove this

File: src/main/java/com/dev/cinema/controller/OrderController.java [hunk 4]
```diff
@@ -34,17 +35,26 @@ public OrderController(OrderService orderService,
...
-    public OrderResponseDto completeOrder(@RequestParam Long userId) {
+    public OrderResponseDto completeOrder(Authentication authentication) {
+        UserDetails details = (UserDetails) authentication.getPrincipal();
+        String email = details.getUsername();
         return orderMapper.toDtoFromObject(orderService
-                .completeOrder(shoppingCartService.getByUser(userService.getById(userId))));
+                .completeOrder(shoppingCartService
+                        .getByUser(userService
+                        .findByEmail(email)
+                        .orElseThrow(() -> new RuntimeException("Incorrect login or password")))));
```
Comment (vicire) [inline src/main/java/com/dev/cinema/controller/OrderController.java]: same here

File: src/main/java/com/dev/cinema/util/PasswordValidator.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    public boolean isValid(Object user, ConstraintValidatorContext constraintValidatorContext) {
+        Object fieldValue = new BeanWrapperImpl(user)
+                .getPropertyValue(field);
+        Object fieldMatchValue = new BeanWrapperImpl(user)
+                .getPropertyValue(fieldMatch);
+
+        if (fieldValue != null) {
+            return fieldValue.equals(fieldMatchValue);
+        } else {
+            return fieldMatchValue == null;
```
Comment (vicire) [inline src/main/java/com/dev/cinema/util/PasswordValidator.java]: at this point if fieldValue == null and fieldMatchValue == null you will register your user

**Review by vicire — CHANGES_REQUESTED** [review]
You need to start using `passwordEncoder` in UserService -> add() for password hashing.

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -19,10 +18,6 @@ public UserServiceImpl(UserDao userDao) {
 
     @Override
     public User add(User user) {
-        byte [] salt = HashUtil.getSalt();
-        user.setPassword(HashUtil.getHashPassword(user
-                .getPassword(), salt));
-        user.setSalt(salt);
```
Comment (vicire) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java:25]: You need to save your encode password in db

**Review by vicire — CHANGES_REQUESTED** [review]

**Review by vicire — APPROVED** [review]

File: src/main/java/com/dev/cinema/config/SecurityConfig.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+@EnableWebSecurity
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+    @Autowired
+    public void configurationGlobal(AuthenticationManagerBuilder auth) throws Exception {
+        auth.inMemoryAuthentication().passwordEncoder(getEncoder()).withUser("bob")
+                .password(getEncoder().encode("1234")).roles("USER");
+    }
+
+    protected void configure(HttpSecurity http) throws Exception {
+        http.authorizeRequests().anyRequest().authenticated()
```
Comment (boroda4436) [inline src/main/java/com/dev/cinema/config/SecurityConfig.java]: better to write each operation on a new line
```suggestion
        http.authorizeRequests()
        .anyRequest().authenticated()
        .and()
        .formLogin().permitAll()
        .and()
        .httpBasic()
```

Comment (boroda4436) [reply src/main/java/com/dev/cinema/config/SecurityConfig.java]: BTW, have you tested your code using postman?

File: src/main/java/com/dev/cinema/controller/ShoppingCartController.java [hunk 8]
```diff
@@ -34,15 +36,21 @@ public ShoppingCartController(ShoppingCartService shoppingCartService,
     }
 
     @PostMapping
-    public void addSession(@RequestParam Long userId, @RequestParam Long movieSessionId) {
-        User user = userService.getById(userId);
+    public void addSession(Authentication authentication, @RequestParam Long movieSessionId) {
+        UserDetails details = (UserDetails) authentication.getPrincipal();
+        String email = details.getUsername();
+        User user = userService.findByEmail(email).orElseThrow(()
+                -> new RuntimeException("Incorrect email or password"));
```
Comment (boroda4436) [inline src/main/java/com/dev/cinema/controller/ShoppingCartController.java]: let's just call `get()` method. NoSuchElementException could be handled in GlobalExceptionHandler (this is not a part of HW)

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/config/SecurityConfig.java (hunk 7 above)
Comment (AntonPanchishko) [reply src/main/java/com/dev/cinema/config/SecurityConfig.java:21]: yeap

File: src/main/java/com/dev/cinema/controller/ShoppingCartController.java (hunk 8 above)
Comment (boroda4436) [reply src/main/java/com/dev/cinema/controller/ShoppingCartController.java]: @AntonPanchishko And here?

**Review by boroda4436 — APPROVED** [review]

**Pull request merged**